### PR TITLE
O11Y-5250: Add retry

### PR DESCRIPTION
**Pull request by changliu-wk**
## Which problem is this PR solving?

When call [client.post](https://github.com/Workiva/opentelemetry-dart/blob/master/lib/src/sdk/trace/exporters/collector_exporter.dart#L54) collector_exporter didn't check if the response is succeeded or not, this PR would like to add retry effort when the response is retryable. 

## How Has This Been Tested?

Please describe the tests that you ran to verify your change. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist:

- [x] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/sdk/trace/exporters/collector_exporter.dart [hunk 1]
```diff
@@ -38,23 +38,43 @@ class CollectorExporter implements sdk.SpanExporter {
       return;
     }
 
-    unawaited(_send(uri, spans));
+    await _send(uri, spans);
```
Comment (blakeroberts-wk) [inline lib/src/sdk/trace/exporters/collector_exporter.dart:41]: What is the purpose of this change?

File: lib/src/sdk/trace/exporters/collector_exporter.dart [hunk 2]
```diff
@@ -38,23 +38,43 @@ class CollectorExporter implements sdk.SpanExporter {
...
-          resourceSpans: _spansToProtobuf(spans));
-      final headers = {'Content-Type': 'application/x-protobuf'}
-        ..addAll(this.headers);
-
-      await client.post(uri, body: body.writeToBuffer(), headers: headers);
-    } catch (e) {
-      _log.warning('Failed to export ${spans.length} spans.', e);
+    const maxRetries = 3;
+    var retries = 0;
+    const valid_retry_codes = [429, 408, 500, 502, 503, 504];
```
Comment (blakeroberts-wk) [inline lib/src/sdk/trace/exporters/collector_exporter.dart:50]: Here is the otel spec on retryable status codes: https://opentelemetry.io/docs/specs/otlp/#failures-1

File: lib/src/sdk/trace/exporters/collector_exporter.dart (hunk 1 above)
Comment (changliu-wk) [reply lib/src/sdk/trace/exporters/collector_exporter.dart:41]: Letting the test case could pass.

File: lib/src/sdk/trace/exporters/collector_exporter.dart [hunk 3]
```diff
@@ -38,23 +38,43 @@ class CollectorExporter implements sdk.SpanExporter {
...
+        _log.warning('Failed to export ${spans.length} spans. '
+            'HTTP status code: ${response.statusCode}');
+        // If the response is not a valid retry code, do not retry
+        if (!valid_retry_codes.contains(response.statusCode)) {
+          return;
+        }
+      } catch (e) {
+        _log.warning('Failed to export ${spans.length} spans. $e');
+      }
+      await Future.delayed(Duration(seconds: retries));
```
Comment (blakeroberts-wk) [inline lib/src/sdk/trace/exporters/collector_exporter.dart:74]: We should use exponential backoff. We should probably add jitter too, though there's implicit disagreement in the spec:
- https://opentelemetry.io/docs/specs/otel/protocol/exporter/#retry
- https://opentelemetry.io/docs/specs/otlp/#otlphttp-throttling

Both say exponential backoff, only one says jitter

File: lib/src/sdk/trace/exporters/collector_exporter.dart (hunk 1 above)
Comment (blakeroberts-wk) [reply lib/src/sdk/trace/exporters/collector_exporter.dart:41]: I believe this is a breaking change because we'd need to update the return type of the abstract class this method overrides

Comment (changliu-wk) [reply lib/src/sdk/trace/exporters/collector_exporter.dart]: Reverted the change back and found a way for the test cases.

File: lib/src/sdk/trace/exporters/collector_exporter.dart (hunk 2 above)
Comment (changliu-wk) [reply lib/src/sdk/trace/exporters/collector_exporter.dart]: Thanks for pointing it out.

File: lib/src/sdk/trace/exporters/collector_exporter.dart (hunk 3 above)
Comment (changliu-wk) [reply lib/src/sdk/trace/exporters/collector_exporter.dart]: Added jitter in.

**Review by michaelyeager-wf — APPROVED** [review]

File: lib/src/sdk/trace/exporters/collector_exporter.dart [hunk 4]
```diff
@@ -45,16 +46,47 @@ class CollectorExporter implements sdk.SpanExporter {
...
+      // Exponential backoff with jitter
+      final delay = calculateJitteredDelay(retries, Duration(seconds: 1));
+      await Future.delayed(delay);
     }
+    _log.severe(
+        'Failed to export ${spans.length} spans after $maxRetries retries');
+  }
+
+  Duration calculateJitteredDelay(int retries, Duration baseDelay) {
+    final random = Random();
```
Comment (blakeroberts-wk) [inline lib/src/sdk/trace/exporters/collector_exporter.dart:86]: Can we make this a class member variable?

File: lib/src/sdk/trace/exporters/collector_exporter.dart [hunk 5]
```diff
@@ -45,16 +46,47 @@ class CollectorExporter implements sdk.SpanExporter {
...
+    final jitter = random.nextDouble() * baseDelay.inMilliseconds;
+    return Duration(
+        milliseconds: baseDelay.inMilliseconds + jitter.toInt() * retries);
```
Comment (blakeroberts-wk) [inline lib/src/sdk/trace/exporters/collector_exporter.dart:89]: We should use exponential delay, something like this:
```suggestion
    final delay = baseDelay.inMilliseconds * pow(2, retries)
    final jitter = random.nextDouble() * delay;
    return Duration(milliseconds: (delay + jitter).toInt());
```

File: lib/src/sdk/trace/exporters/collector_exporter.dart [hunk 6]
```diff
@@ -45,16 +46,47 @@ class CollectorExporter implements sdk.SpanExporter {
...
+        // If the response is not a valid retry code, do not retry
+        if (!valid_retry_codes.contains(response.statusCode)) {
+          return;
+        }
+      } catch (e) {
+        _log.warning('Failed to export ${spans.length} spans. $e');
+        return;
+      }
+      // Exponential backoff with jitter
+      final delay = calculateJitteredDelay(retries, Duration(seconds: 1));
```
Comment (blakeroberts-wk) [inline lib/src/sdk/trace/exporters/collector_exporter.dart:78]: I think it's reasonable to start with something smaller like 100 milliseconds

File: lib/src/sdk/trace/exporters/collector_exporter.dart (hunk 5 above)
Comment (blakeroberts-wk) [reply lib/src/sdk/trace/exporters/collector_exporter.dart:89]: This assumes the first retry will have "retries" set to zero, not one

**Review by blakeroberts-wk — APPROVED** [review]

**Comment by changliu-wk** [comment]
@Workiva/release-management-p

**Comment by changliu-wk** [comment]
QA +1

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM
